src/database/migrations/20230803014224_CreateAuditTrailsTableAndMigrateData.ts

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable camelcase */
+import { Knex } from 'knex';
+
+import {
+  AuditTrailDTO,
+  AuditTrailEvents,
+  SharingAction,
+  UpdateSharingPermissionEventData,
+} from '../../api/entities/AuditTrail';
+
+type SharingAuditTrail = {
+  action: SharingAction;
+  siteIds?: string;
+  sharingParticipantSiteId?: number;
+  userId: number;
+  userEmail: string;
+  participantId: number;
+  proceed: boolean;
+  created_at: Date;
+  updated_at: Date;
+};
+
+function mapSharingAuditTrailToAuditTrails(
+  sharingAuditTrail: SharingAuditTrail
+): Omit<AuditTrailDTO, 'id'> & { created_at: Date; updated_at: Date } {
+  const eventData: UpdateSharingPermissionEventData = {
+    action: sharingAuditTrail.action,
+    siteId: 0, // Legacy Data will miss siteId
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    sharingPermissions: JSON.parse(`[${sharingAuditTrail.siteIds!}]`),
+  };
+
+  return {
+    userId: sharingAuditTrail.userId,
+    userEmail: sharingAuditTrail.userEmail,
+    participantId: sharingAuditTrail.participantId,
+    event: AuditTrailEvents.UpdateSharingPermissions,
+    eventData,
+    proceed: sharingAuditTrail.proceed,
+    created_at: sharingAuditTrail.created_at,
+    updated_at: sharingAuditTrail.updated_at,
+  };
+}
+
+async function migrateSharingAuditTrails(knex: Knex) {
+  const groupByColumns = [
+    'userId',
+    'userEmail',
+    'action',
+    'participantId',
+    'proceed',
+    'updated_at',
+    'created_at',
+  ];
+  const sharingAuditTrails = await knex('sharingAuditTrails')
+    .select(knex.raw(`string_agg(sharingParticipantSiteId, ',') siteIds`), ...groupByColumns)
+    .groupBy(groupByColumns);
+  const auditTrails = sharingAuditTrails.map(mapSharingAuditTrailToAuditTrails);
+  await knex('auditTrails').insert(auditTrails);
+}
+
+function mapAuditTrailsToSharingAuditTrails(
+  auditTrail: AuditTrailDTO & { created_at: Date; updated_at: Date }
+): SharingAuditTrail[] {
+  const eventData = JSON.parse(auditTrail.eventData as string) as UpdateSharingPermissionEventData;
```

Comment:
Feel a bit annoying if we always need to do JSON.parse on auditTrail.eventData. Open to suggestions